tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1173,183 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
 {
+    int negative_one = -1;
     for (uint32_t i = 0; i < count; i++) {
         const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
 
-        if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
-            return 1;
+        if (s2n_constant_time_equals(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            return i;
         }
     }
 
-    return 0;
+    return negative_one;
+}
+
+static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+{
+    int server_index = s2n_wire_ciphers_has_server_cipher_at(match, wire, count, cipher_suite_len);
+    if (server_index < 0) {
+        return 0;
+    }
+    return 1;
+}
+
+/* Checks that a mutually-supported cipher suite can be used in this connection */
+static bool s2n_cipher_suite_match_is_valid(struct s2n_connection* conn, struct s2n_cipher_suite* potential_match) {
+        /* Never use TLS1.3 ciphers on a pre-TLS1.3 connection, and vice versa */
+        if ((conn->actual_protocol_version >= S2N_TLS13) != (potential_match->minimum_required_tls_version >= S2N_TLS13)) {
+            return false;
+        }
+
+        /* If connection is for SSLv3, use SSLv3 version of suites */
+        if (conn->client_protocol_version == S2N_SSLv3) {
+            potential_match = potential_match->sslv3_cipher_suite;
+        }
+
+        /* Skip the suite if we don't have an available implementation */
+        if (!potential_match->available) {
+            return false;
+        }
+
+        /* Make sure the cipher is valid for available certs */
+        if (s2n_is_cipher_suite_valid_for_auth(conn, potential_match) != S2N_SUCCESS) {
+            return false;
+        }
+
+        /* TLS 1.3 does not include key exchange in cipher suites */
+        if (potential_match->minimum_required_tls_version < S2N_TLS13) {
+            /* If the kex is not supported continue to the next candidate */
+            bool kex_supported = false;
+            if (s2n_result_is_error(s2n_kex_supported(potential_match, conn, &kex_supported))) {
+                return false;
+            }
+
+            if (!kex_supported) {
+                return false;
+            }
+
+            /* If the kex is not configured correctly continue to the next candidate */
+            if (s2n_result_is_error(s2n_configure_kex(potential_match, conn))) {
+                return false;
+            }
+        }
+
+        /**
+        *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
+        *# The server MUST ensure that it selects a compatible PSK
+        *# (if any) and cipher suite.
+        **/
+        if (conn->psk_params.chosen_psk != NULL) {
+            if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+                return false;
+            }
+        }
+
+        return true;
+}
+
+/**
+ *  TODO: Add basic validation for the cipher preference list
```

Comment:
When we initialize the security policies we preform basic validation on them: https://github.com/aws/s2n-tls/blob/main/tls/s2n_security_policies.c#L911. So that validation logic would live in that file, rather than in this function, which is good, because validation logic would make this code even more confusing :). 
I would remove this TODO and instead make an issue in github for this feature and add it to the list there(like this: https://github.com/aws/s2n-tls/issues/3442)

Reply:
Ok